### PR TITLE
feat(server): report slaves ip and listening port in INFO REPLICATION

### PR DESCRIPTION
**Pull request by ashotland**
See #706 for more context

Signed-off-by: ashotland <ashotland@gmail.com>

### CODE REVIEW

File: src/server/dflycmd.cc [hunk 1]
```diff
@@ -296,8 +297,8 @@ void DflyCmd::Flow(CmdArgList args, ConnectionContext* cntx) {
 
   // Set meta info on connection.
   cntx->owner()->SetName(absl::StrCat("repl_flow_", sync_id));
-  cntx->conn_state.repl_session_id = sync_id;
-  cntx->conn_state.repl_flow_id = flow_id;
+  cntx->conn_state.replicaiton_info.repl_session_id = sync_id;
```
Comment (ashotland) [inline src/server/dflycmd.cc:300]: non functional change

File: src/server/dflycmd.cc [hunk 2]
```diff
@@ -296,8 +297,8 @@ void DflyCmd::Flow(CmdArgList args, ConnectionContext* cntx) {
...
+  cntx->conn_state.replicaiton_info.repl_flow_id = flow_id;
```
Comment (ashotland) [inline src/server/dflycmd.cc:301]: non functional change

File: src/server/dflycmd.cc [hunk 3]
```diff
@@ -499,7 +501,7 @@ uint32_t DflyCmd::CreateSyncSession(ConnectionContext* cntx) {
 }
 
 void DflyCmd::OnClose(ConnectionContext* cntx) {
-  unsigned session_id = cntx->conn_state.repl_session_id;
+  unsigned session_id = cntx->conn_state.replicaiton_info.repl_session_id;
```
Comment (ashotland) [inline src/server/dflycmd.cc:504]: non functional change

File: src/server/server_family.cc [hunk 4]
```diff
@@ -1549,7 +1555,7 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
         cntx->owner()->SetName(absl::StrCat("repl_ctrl_", sid));
 
         string sync_id = absl::StrCat("SYNC", sid);
-        cntx->conn_state.repl_session_id = sid;
+        cntx->conn_state.replicaiton_info.repl_session_id = sid;
```
Comment (ashotland) [inline src/server/server_family.cc:1558]: non functional change

File: src/server/conn_context.h [hunk 5]
```diff
@@ -75,6 +75,14 @@ struct ConnectionState {
     util::fibers_ext::BlockingCounter borrow_token{0};
   };
 
+  struct ReplicationInfo {
```
Comment (ashotland) [inline src/server/conn_context.h:78]: Tiny refactor of the replication related fields into a struct

File: src/server/server_family.cc [hunk 6]
```diff
@@ -1563,6 +1569,10 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
         (*cntx)->SendLong(shard_set->pool()->size());
         return;
       }
+    } else if (cmd == "LISTENING-PORT") {
+      uint32_t replica_listening_port;
+      absl::SimpleAtoi(arg, &replica_listening_port);
```
Comment (adiholden) [inline src/server/server_family.cc:1574]: check return status

File: src/server/server_family.cc [hunk 7]
```diff
@@ -1362,6 +1362,12 @@ void ServerFamily::Info(CmdArgList args, ConnectionContext* cntx) {
     if (etl.is_master) {
       append("role", "master");
       append("connected_slaves", m.conn_stats.num_replicas);
+      auto replicas = dfly_cmd_->GetReplicasRoleInfo();
+      for (auto i = 0; i < replicas.size(); i++) {
+        auto& r = replicas[i];
+        // e.g. slave0:ip=172.19.0.3,port=6379
+        append(StrCat("slave", i), StrCat("ip=", r.address, ",port=", r.listening_port));
```
Comment (adiholden) [inline src/server/server_family.cc:1369]: can you write replica instead of slave?

File: src/server/dflycmd.h [hunk 8]
```diff
@@ -101,21 +101,24 @@ class DflyCmd {
...
     Context cntx;
 
     std::vector<FlowInfo> flows;
     ::boost::fibers::mutex mu;  // See top of header for locking levels.
   };
   struct ReplicaRoleInfo {
-    ReplicaRoleInfo(std::string address, SyncState sync_state);
+    ReplicaRoleInfo(std::string address, uint32_t listening_port, SyncState sync_state);
     std::string address;
+    uint_fast32_t listening_port;
```
Comment (adiholden) [inline src/server/dflycmd.h:121]: why uint_fast32_t?

File: src/server/dflycmd.h [hunk 9]
```diff
@@ -101,21 +101,24 @@ class DflyCmd {
 
   // Stores information related to a single replica.
   struct ReplicaInfo {
-    ReplicaInfo(unsigned flow_count, std::string address, Context::ErrHandler err_handler)
-        : state{SyncState::PREPARATION}, address{address}, cntx{std::move(err_handler)},
-          flows{flow_count} {
+    ReplicaInfo(unsigned flow_count, std::string address, int32_t listening_port,
```
Comment (adiholden) [inline src/server/dflycmd.h:104]: int32_t  ->  uint32_t

File: src/server/server_family.cc (hunk 7 above)
Comment (ashotland) [reply src/server/server_family.cc:1369]: Nope, redis-sentinel parses it as 'slave'

**Comment by ashotland** [comment]
Thanks @adiholden - all comments addressed

File: src/server/server_family.cc [hunk 10]
```diff
@@ -1563,6 +1569,12 @@ void ServerFamily::ReplConf(CmdArgList args, ConnectionContext* cntx) {
         (*cntx)->SendLong(shard_set->pool()->size());
         return;
       }
+    } else if (cmd == "LISTENING-PORT") {
+      uint32_t replica_listening_port;
+      if (!absl::SimpleAtoi(arg, &replica_listening_port)) {
+        (*cntx)->SendError(kInvalidIntErr);
+      }
```
Comment (adiholden) [inline src/server/server_family.cc:1576]: return

**Comment by adiholden** [comment]
@ashotland it will be better if you push another commit with fixes (not edit the commit) for next times so we can review only the new changes that were pushed

File: src/server/server_family.cc (hunk 10 above)
Comment (ashotland) [reply src/server/server_family.cc:1577]: @adiholden added thanks

**Review by adiholden — APPROVED** [review]

**Pull request merged**